src/components/cuda_pw/cupti_component.c

```diff
@@ -0,0 +1,95 @@
+#include <papi.h>
+#include <papi_internal.h>
+#include <papi_vector.h>
+
+#include "debug_comp.h"
+#include "api_cuda.h"
+#include "common_defs.h"
+#include "cuda_common.h"
+#include "cupti_component.h"
+
+#if API_PERFWORKS
+#include "cupti_profiler.h"
+#elif API_EVENTS
+#include "cupti_events.h"
+#endif
+
+extern papi_vector_t _cuda_pw_vector;
+
+int cupti_shutdown(void)
+{
+    return PAPI_OK;
+}
+
+int cupti_init(const char **pdisabled_reason)
+{
+    int retval;
+#if API_PERFWORKS
+    retval = cupti_profiler_init(pdisabled_reason);
+#elif API_EVENTS
+    retval = cupti_events_init(pdisabled_reason);
+#else
+    *pdisabled_reason = "CUDA not found.";
+    retval = PAPI_ECMP;
+#endif
+    if (retval != PAPI_OK)
+    {
+        strcpy(_cuda_pw_vector.cmp_info.disabled_reason, *pdisabled_reason);
+        goto fn_fail;
+    }
+    _cuda_pw_vector.cmp_info.disabled = PAPI_OK;
+    return PAPI_OK;
+fn_fail:
+    return PAPI_ECMP;
+}
+
+int cupti_init_cuctx_arr(void ** pcuda_context)
+{
+    return init_CUcontext_array(pcuda_context);
+}
+
+int cupti_control_create(struct event_name_list_s * event_names, int event_count, int *evt_ids, void ** pcupti_ctl, void **pcu_ctx)
+{
+#if API_PERFWORKS
+    int res = cupti_profiler_control_create(event_names, event_count, evt_ids, pcupti_ctl, pcu_ctx);
+    return res;
+#endif
+}
+
+int cupti_control_destroy(void **pcupti_ctl)
+{
+#if API_PERFWORKS
+    int res = cupti_profiler_control_destroy(pcupti_ctl);
+    return res;
+#endif
+}
```

Comment:
Too fine-grained interface. The cupti layer should only expose `cupti_ctx_open` (or `cupti_ctl_open`). If you really need to separate initialization you can have a `cupti_pre_init` and `cupti_post_init` instead of `cupti_init` only.